% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/getPersistence.R
\name{getPersistence}
\alias{getPersistence}
\title{Computes Persistence Scores For a Single Time-Series Across a Single Lag}
\usage{
getPersistence(timeSeries, lag, laplacian = T)
}
\arguments{
\item{timeSeries}{a \code{vector} of \code{numeric} time-series expression values.}

\item{lag}{a \code{numeric} specifying the Lag to use for in the 3-D time delayed embedding.}

\item{laplacian}{a \code{logical} scalar. Should the Laplacian Eigenmaps be used for dimensionality reduction? Default \code{TRUE}.}
}
\value{
the max persistence score at the specified lag, returns 0 if no persistence score exists.
}
\description{
Takes a \code{vector} of numeric gene expression over time and computes the persistence score.
The specified lag is used to transform the expression into a 3-D embedded space via time-delay embedding.
A non-linear dimension reduction technique (laplacian eigenmaps) is used to transfrom the 3-D embedding to a 2-D embedding.
Finally, the persistence score of the 2-D embedding is calculated via persistence homology.
Returns the Max persistence score, returns 0 if no persistence score exists.
For more details see TimeCycle's vignette:
\code{vignette("TimeCycle")}.
}
\references{
{
   \itemize{
     \item Wadhwa RR, Williamson DFK, Dhawan A, Scott JG. (2018). "TDAstats: R pipeline for computing persistent homology in topological data analysis." \emph{Journal of Open Source Software}. 2018; 3(28): 860. doi:\href{https://doi.org/10.21105/joss.00860}{[10.21105/joss.00860]}
     \item Bauer U. (2019). "Ripser: Efficient computation of Vietoris-Rips persistence barcodes." \emph{arXiv}: 1908.02518.
   }
}
}
\seealso{
\itemize{
     \item \code{\link[TDAstats]{calculate_homology}} for Persistence Homology calculation.
     \item \code{\link{buildTakens_ndim}} for for generating time-delay embedding.
     \item \code{\link{computeLaplacianEmbedding}} for 3-D to 2-D laplacian eigenmaps dimension reduction.
     \item \code{\link{computePersistence}} for use parallelized function for a \code{data.frame} of gene expression.
   }
}
