<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Computes Persistence Scores For a Data.Frame of Time-Series Across Multiple Lags — computePersistence • TimeCycle</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Computes Persistence Scores For a Data.Frame of Time-Series Across Multiple Lags — computePersistence" />
<meta property="og:description" content="Takes a data.frame of numeric gene expression over time (genes X ZT times) and computes the persistence score using getPersistence.
For a given gene, each lag (min to max) is used to transform the expression into a 3-D embedded space via time-delay embedding.
A non-linear dimension reduction technique (laplacian eigenmaps) is used to transfrom the 3-D embedding to a 2-D embedding.
Finally, the persistence score of the 2-D embedding is calculated via persistence homology.
The median persistence score across all lags (min to max) for each gene is returned as a numeric vector.
For more details see TimeCycle's vignette:
vignette(&quot;TimeCycle&quot;)." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TimeCycle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/TimeCycle.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    News
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Releases</li>
    <li class="dropdown-header">Version 1.0.0</li>
    <li class="divider"></li>
    <li>
      <a href="../news/index.html">Changelog</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/nesscoder/TimeCycle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Computes Persistence Scores For a Data.Frame of Time-Series Across Multiple Lags</h1>
    <small class="dont-index">Source: <a href='https://github.com/nesscoder/TimeCycle/blob/master/R/computePersistence.R'><code>R/computePersistence.R</code></a></small>
    <div class="hidden name"><code>computePersistence.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Takes a <code>data.frame</code> of numeric gene expression over time (genes X ZT times) and computes the persistence score using <code><a href='getPersistence.html'>getPersistence</a></code>.
For a given gene, each lag (min to max) is used to transform the expression into a 3-D embedded space via time-delay embedding.
A non-linear dimension reduction technique (laplacian eigenmaps) is used to transfrom the 3-D embedding to a 2-D embedding.
Finally, the persistence score of the 2-D embedding is calculated via persistence homology.
The median persistence score across all lags (min to max) for each gene is returned as a numeric vector.
For more details see TimeCycle's vignette:
<code><a href='../articles/TimeCycle.html'>vignette("TimeCycle")</a></code>.</p>
    </div>

    <pre class="usage"><span class='fu'>computePersistence</span><span class='op'>(</span>
  <span class='va'>data</span>,
  minLag <span class='op'>=</span> <span class='fl'>2</span>,
  maxLag <span class='op'>=</span> <span class='fl'>5</span>,
  cores <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>-</span> <span class='fl'>2</span>,
  laplacian <span class='op'>=</span> <span class='cn'>T</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data</th>
      <td><p>a <code>data.frame</code> of <code>numeric</code> gene expression over time (row = genes <em>x</em> col = ZT times).</p></td>
    </tr>
    <tr>
      <th>minLag</th>
      <td><p>a <code>numeric</code> specifying the min lag to check in the 3-D embedding. Default is <code>2</code>.</p></td>
    </tr>
    <tr>
      <th>maxLag</th>
      <td><p>a <code>numeric</code> specifying the max lag to check in the 3-D embedding. Default is <code>5</code>.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>a <code>numeric</code> specifying the number of parallel cores to use. Default number of cores is <code>parallel::detectedCores() - 2</code>.</p></td>
    </tr>
    <tr>
      <th>laplacian</th>
      <td><p>a <code>logical</code> scalar. Should the Laplacian Eigenmaps be used for dimensionality reduction? Default <code>TRUE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>vector</code> of the median persistence score across lags (minLag to maxLag) for each gene in data</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><ul>
<li><p>Fasy, Brittany &amp; Kim, Jisu &amp; Lecci, Fabrizio &amp; Maria, Clément. (2014). "Introduction to the R package TDA".</p></li>
<li><p>Maria C (2014). "GUDHI, Simplicial Complexes and Persistent Homology Packages." <a href=' https://project.inria.fr/gudhi/software/ '> https://project.inria.fr/gudhi/software/ </a>.</p></li>
<li><p>Morozov D (2007). "Dionysus, a C++ library for computing persistent homology". <a href=' http://www.mrzv.org/software/dionysus/ '> http://www.mrzv.org/software/dionysus/ </a></p></li>
<li><p>Edelsbrunner H, Harer J (2010). "Computational topology: an introduction." American Mathematical Society.</p></li>
<li><p>Fasy B, Lecci F, Rinaldo A, Wasserman L, Balakrishnan S, Singh A (2013). "Statistical Inference For Persistent Homology." (arXiv:1303.7117). Annals of Statistics.</p></li>
</ul></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='https://rdrr.io/pkg/TDA/man/ripsDiag.html'>ripsDiag</a></code> for Persistence Homology calculation.</p></li>
<li><p><code><a href='buildTakens_ndim.html'>buildTakens_ndim</a></code> for for generating time-delay embedding.</p></li>
<li><p><code><a href='computeLaplacianEmbedding.html'>computeLaplacianEmbedding</a></code> for 3-D to 2-D laplacian eigenmaps dimension reduction.</p></li>
<li><p><code><a href='getPersistence.html'>getPersistence</a></code> for use on a single gene expression time-series.</p></li>
</ul>
</div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://sites.northwestern.edu/elannesscohn/'>Elan Ness-Cohn</a>, <a href='https://sites.northwestern.edu/braunlab/'>Rosemary Braun</a>, <a href='https://www.quantitativebiology.northwestern.edu/'><img src='reference/figures/QBio.png' height=40></a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


